test/functional/sqlsrv/sqlsrv_ae_insert_sqltype_datetime_precision.phpt

```diff
@@ -0,0 +1,181 @@
+    if ($precision == 7) {
+        $dtstr = substr($dtstr, 0, -1);
+        if (!is_null($dtarr)) {
+            $dtarr[1] = substr($dtarr[1], 0, -1);
+        }
+    }
+    if (strpos($dtobj_date, $dtstr) !== false) {
+        return true;
+    }
+    if ($dataType == "datetimeoffset") {
+        if (strpos($dtobj_date, $dtarr[0]) !== false && strpos($dtobj_date, $dtarr[1]) !== false && strpos($dtobj_timezone, $dtarr[2]) !== false) {
+            return true;
+        }
+    }
+    return false;
+}
+
+$dataTypes = array("datetime2", "datetimeoffset", "time");
+$precisions = array(0, 1, 2, 4, 7);
+$inputValuesInit = array("datetime2" => array("0001-01-01 00:00:00", "9999-12-31 23:59:59"),
+                         "datetimeoffset" => array("0001-01-01 00:00:00 -14:00", "9999-12-31 23:59:59 +14:00"),
+                         "time" => array("00:00:00", "23:59:59"));
+
+$conn = AE\connect();
+foreach($dataTypes as $dataType) {
+    foreach($precisions as $m) {
+        // add $m number of decimal digits to the some input values
+        $inputValues[0] = $inputValuesInit[$dataType][0];
+        $inputValues[1] = $inputValuesInit[$dataType][1];
+        if ($m != 0) {
+            if ($dataType == "datetime2") {
+                $inputValues[1] .= "." . str_repeat("4", $m);
+            } else if ($dataType == "datetimeoffset") {
+                $dtoffsetPieces = explode(" ", $inputValues[1]);
+                $inputValues[1] = $dtoffsetPieces[0] . " " . $dtoffsetPieces[1] . "." . str_repeat("4", $m) . " " . $dtoffsetPieces[2];
+            } else if ($dataType == "time") {
+                $inputValues[0] .= "." . str_repeat("0", $m);
+                $inputValues[1] .= "." . str_repeat("4", $m);
+            }
+        }
+        $typeFull = "$dataType($m)";
+        echo "\nTesting $typeFull:\n";
+    
+        // create table containing datetime2(m), datetimeoffset(m) or time(m) columns
+        $tbname = "test_" . $dataType . $m;
+        $colMetaArr = array(new AE\ColumnMeta($typeFull, "c_det"), new AE\ColumnMeta($typeFull, "c_rand", null, false));
+        AE\createTable($conn, $tbname, $colMetaArr);
+        
+        // insert by specifying the corresponding SQLSRV_SQLTYPE
+        $sqlType = "SQLSRV_SQLTYPE_" . strtoupper($dataType);
+        $inputs = array(new AE\BindParamOption($inputValues[0], null, null, $sqlType),
+                        new AE\BindParamOption($inputValues[1], null, null, $sqlType));
+        $r;
+        $stmt = AE\insertRow($conn, $tbname, array("c_det" => $inputs[0], "c_rand" => $inputs[1]), $r, AE\INSERT_PREPARE_PARAMS);
+        
+        // check the case when the column precision (m) is less than 7
+        // with AE: should not work
+        // with AE: should work
+        if ($m < 7) {
+            if (AE\isColEncrypted()) {
+                if ($r !== false) {
+                    echo "AE: Conversion from $sqlType to $typeFull should not be supported\n";
+                } else {
```

Comment:
again, change the error to "should have failed" instead -- please change that for other tests too